crates/core/src/crud_vtab.rs

```diff
@@ -176,41 +182,90 @@ impl VirtualTable {
         self.current_tx = Some(ActiveCrudTransaction {
             tx_id,
             mode: if self.is_simple {
-                CrudTransactionMode::Simple {
-                    // language=SQLite
-                    stmt: db.prepare_v3("INSERT INTO ps_crud(tx_id, data) VALUES (?, ?)", 0)?,
-                    // language=SQLite
-                    set_updated_rows: db.prepare_v3(
-                        "INSERT OR IGNORE INTO ps_updated_rows(row_type, row_id) VALUES(?, ?)",
-                        0,
-                    )?,
-                    update_local_bucket: db.prepare_v3(formatcp!("INSERT OR REPLACE INTO ps_buckets(name, last_op, target_op) VALUES('$local', 0, {MAX_OP_ID})"), 0)?,
-                }
+                CrudTransactionMode::Simple(Default::default())
             } else {
-                const SQL: &str = formatcp!(
-                    "\
+                CrudTransactionMode::Manual(Default::default())
+            },
+        });
+
+        Ok(())
+    }
+
+    fn end_transaction(&mut self) -> Result<(), SQLiteError> {
+        let tx = self.current_tx.take();
+        if let Some(tx) = tx {
+            let needs_local_bucket_update = match tx.mode {
+                CrudTransactionMode::Manual { .. } => {
+                    // In manual mode, users need to update the $local bucket themselves.
+                    false
+                }
+                CrudTransactionMode::Simple(simple) => simple.had_writes,
+            };
+
+            if needs_local_bucket_update {
+                self.db.exec_safe(formatcp!("INSERT OR REPLACE INTO ps_buckets(name, last_op, target_op) VALUES('$local', 0, {MAX_OP_ID})"))?;
```

Comment:
Is it safe to do this inside the `xCommit` callback? As in, are we sure that this will be called before the sync/commit process is started on the underlying VFS? I can't find anything mentioning the order of operations in the [vtab docs](https://www.sqlite.org/vtab.html), so not sure how to check that.

Reply:
Good catch! I think the `xSync` method would be a better fit for this, and `xCommit` definitely isn't safe. However, we can also just run the statement once on the first write which sounds like the easiest option.